### ISSUE TITLE
[Feature] Enum InvalidCastException analysis

### ISSUE THREAD

**Issue by jzabroski** [DotNetAnalyzers/ReflectionAnalyzers#215]
Placeholder to fully describe positive and negative test cases.

```csharp
public enum ABEnum { A = 0, B = 1}
public class Test
{
  public void TestAB()
  {
    foreach (var item in Enum.GetValues(typeof(ABEnum)))
    {
      Console.WriteLine(((Enum)item).GetDisplayName()); // Feature logged in: https://github.com/DotNetAnalyzers/ReflectionAnalyzers/issues/214
                     // ^^^^^^^^^^^^ bonus: don't raise Possible InvalidCastException if typeof(ABEnum) is an enum defined as an int64 sub-class (default scenario).
    }
  }
}
```

Ideally, the code will recommend the correct cast to the appropriate integer size.  For example, if the lvalue being assigned to is an int, and the enum is a long, it will recommend a down-cast.  If the lvalue being assigned to is a long, and the enum is a short, there will be no loss of precision but it may make sense to warn low severity about data type mismatch.

**Comment by brian-reichle** [comment]
I cant seem to reconcile your description with the sample code.

Is the issue with casting the result of `Enum.GetValues` to the correct type?

```csharp
foreach (var item in Enum.GetValues(typeof(ABEnum)))
{
    Console.WriteLine((ABEnum)item); // ok, we know item is an instance of ABEnum
    Console.WriteLine((Enum)item); // ok, ABEnum inherits from Enum
    Console.WriteLine((int)(ABEnum)item); // ok, there is a conversion cast from ABEnum to int.

    Console.WriteLine((int)item); // not ok, we need to perform the unboxing cast before we can perform the conversion cast.
}
```
or is the issue with value truncation?

```csharp
public enum ABEnum { A = 0, B = 1 }
public enum CDEnum : long { A = 0, B = 1 }

void Foo(ABEnum a, CDEnum c)
{
    var value1 = (int)a; // ok
    var value2 = (long)c; // ok

    var value3 = (int)c; // warn - truncating conversion.
    var value4 = (long)a; // ok - widening conversion.
}
```

**Comment by jzabroski** [comment]
It should be split into two cases. I was just quickly jotting down thoughts as fast and clear as possible while trying to get my tasks done at work.

Specifically, truncation should be a separate issue. There should probably be a third analysis for out of range casts:

```csharp
public enum ABEnum { A = 1, B = 2 }
public bool TestAB() {
  const int c = 3;
  return (ABEnum)c; // out of range value for cast
} 
```

For out of range, it should probably suggest '(typeof(ABEnum).IsEnumDefined(c))`
